bookkeeper-server/src/test/java/org/apache/bookkeeper/client/BookieWriteLedgerTest.java

```diff
@@ -710,6 +711,83 @@ public void testLedgerCreateAdvWithAsyncWritesWithBookieFailures() throws Except
         lh2.close();
     }
 
+    /**
+     * LedgerHandleAdv out of order writers with ensemble changes.
+     * Verify that entry that was written to old ensemble will be
+     * written to new enseble too after ensemble change.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testLedgerHandleAdvOutOfOrderWriteAndFrocedEnsembleChange() throws Exception {
+        // Create a ledger
+        long ledgerId = 0xABCDEF;
+        SyncObj syncObj1 = new SyncObj();
+        ByteBuffer entry;
+        lh = bkc.createLedgerAdv(ledgerId, 3, 3, 3, digestType, ledgerPassword, null);
+        entry = ByteBuffer.allocate(4);
+        // Add entries 0-4
+        for (int i = 0; i < 5; i++) {
+            entry.rewind();
+            entry.putInt(rng.nextInt(maxInt));
+            lh.addEntry(i, entry.array());
+        }
+
+        // Add 10 as Async Entry, which goes to first ensemble
+        ByteBuffer entry1 = ByteBuffer.allocate(4);
+        entry1.putInt(rng.nextInt(maxInt));
+        lh.asyncAddEntry(10, entry1.array(), 0, entry1.capacity(), this, syncObj1);
+
+        Thread.sleep(5000);// wait 5 sec so entry-11 goes to the bookies and gets response.
```

Comment:
can we avoid `Thread.sleep(5000)`? You can inspect ledger handle `lh` to see if pending add op it has received response.

Reply:
You mean to say create a test visible method in ledger handle to peek into pendingAddOps? Or is there something I can get this info from?

Reply:
You can use powemock Whitebox getInternalState in orde to access internal without exposing new method

Reply:
You can use powemock Whitebox getInternalState in orde to access internal without exposing new method